### PR TITLE
Issue 22974: Initial bugfix suggestion.

### LINKED ISSUE #22974: Creature/script: High Priestess Mar'li
**Description:** This boss https://www.wowhead.com/npc=14510/high-priestess-marli must hatch spider eggs https://www.wowhead.com/object=179985/spider-egg that summon a spider.

In the current script the spider is summoned manually:
if (Creature* spider = me->SummonCreature(NPC_SPIDER, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000))

Spell that must be used to hatch eggs: https://woehead.way-of-elendil.fr/?spell=24082

Hatched eggs must respawn after a wipe.

**Branch(es):** 3.3.5

**TC rev. hash/commit:** rev. 813cf761fcf8

### PR DESCRIPTION
**Pull request by kuskmen**
**Changes proposed:**

-  Spiders are now being hatched from the nearby Spider Eggs instead of being spawn.
-  Eggs should not be hatched while in spider form.

**Target branch(es):** 3.3.5

**Issues addressed:** Closes #22974 

**Tests performed:** Built and tested in-game.

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] `_hatchedEggs` container does not preserve the state of real Spider Eggs, once added to the container on the next update of the AI, container is full of pointers with default/corrupted values leading to **access violation exception sometime in the script**. 
- [ ] Spider targeting needs to be fixed and amended regarding spider's guids rather entries otherwise it will not work as expected.

I am opening this PR with the idea of someone to help me out with the known issues since this is my first script and I am unable to continue/have no idea.

Best regards,
kuskmen

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp [hunk 1]
```diff
@@ -77,142 +80,180 @@ enum Misc
 float const DamageIncrease = 35.0f;
 float const DamageDecrease = 100.f / (1.f + DamageIncrease / 100.f) - 100.f;
 
+float const MaxRangeHatch = 100.f;
```
Comment (kuskmen) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: It's constant I came up with, if you know any reliable source I can take from I will be happy to change.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp [hunk 2]
```diff
@@ -51,7 +53,7 @@ enum Events
 {
     EVENT_SPAWN_START_SPIDERS = 1, // Phase 1
     EVENT_POISON_VOLLEY       = 2, // Phase All
-    EVENT_SPAWN_SPIDER        = 3, // Phase All
+    EVENT_HATCH_SPIDER_EGG        = 3, // Phase All
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: align.

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp [hunk 3]
```diff
@@ -77,142 +80,180 @@ enum Misc
...
+                boss_marliAI(Creature* creature) : BossAI(creature, DATA_MARLI) { }
 
-            void JustDied(Unit* /*killer*/) override
-            {
-                _JustDied();
-                Talk(SAY_DEATH);
-            }
+                void Initialize()
+                {
+                    _hatchedEggs = std::set<GameObject*>();
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: is this line needed ?

Comment (kuskmen) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: I think in theory no, but I've read [somewhere](https://drive.google.com/file/d/0Bw08_9WDSwhKVHBRT0dNc0dZQ2c/view) that it is good practice and convention in TC to initialize class members like that.

But I can delete it I am also not a fan.

Comment (Shauren) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: Not needed for types that have a default constructor (set obviously does)

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp [hunk 4]
```diff
@@ -77,142 +80,180 @@ enum Misc
...
 
-                    if (me->HasUnitState(UNIT_STATE_CASTING))
-                        return;
+                    DoMeleeAttackIfReady();
                 }
 
-                DoMeleeAttackIfReady();
-            }
+            private:
+                std::set<GameObject*> _hatchedEggs;
```
Comment (Shauren) [inline src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: never store pointers to other world objects in scripts - if the object despawns after you store it in here but before you use it again, *bad things will happen* (bad things means accessing freed memory and most likely a crash)

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp (hunk 4 above)
Comment (kuskmen) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: Yes, thats exactly what it is happening, but then how am I suppose to store despawned objects in order to spawn them again then?

Comment (mdX7) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: in general you should save their ObjectGuids and get their concrete objects via ObjectAccessor

Comment (Shauren) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: Pretty sure it will be safe to just respawn ALL of them at reset

Comment (Killyana) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_marli.cpp]: Use something like this https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Kalimdor/RuinsOfAhnQiraj/boss_buru.cpp#L77-L86

**Comment by kuskmen** [comment]
Turns out branch name was bugging my dynamic linking scripts cache directory so I had to rename the branch, I will reopen new PR when ready.

**Pull request closed without merging**